Don't always test string claims for Date Time

keegan-caruso opened the pull request:
# Don't always test string claims for Date Time

## Description

Cannot remove APIs in in these internal methods as they are friends

There are serveral claim types that should never
be a datetime value. This changes introduces
a list of claim types that will not be tested if they are a Date Time format. 
This list of claim types is not exhaustive.

Added test data for a test token with more claims
and a benchmark using this data.

Benchmark results:

BenchmarkDotNet v0.13.12, Windows 11 (10.0.22631.3593/23H2/2023Update/SunValley3) (Hyper-V)
Intel Xeon Platinum 8370C CPU 2.80GHz, 1 CPU, 16 logical and 8 physical cores
.NET SDK 8.0.301
  [Host]    : .NET 8.0.6 (8.0.624.26715), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI
  MediumRun : .NET 8.0.6 (8.0.624.26715), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI

Job=MediumRun  MaxAbsoluteError=10.0000 ms  IterationCount=15
LaunchCount=4  WarmupCount=10

| Method                                                                | Mean     | Error    | StdDev   | P90      | P95      | P100     | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|---------------------------------------------------------------------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|--------:|-------:|----------:|------------:|
| JsonWebTokenHandler_ValidateTokenAsync_CreateClaims                   | 45.71 us | 0.565 us | 1.227 us | 47.23 us | 47.35 us | 47.46 us |  1.10 |    0.03 | 0.6104 |  17.63 KB |        1.04 |
| JsonWebTokenHandler_ValidateTokenAsync_CreateClaims_WithOptimizations | 41.50 us | 0.195 us | 0.416 us | 42.06 us | 42.14 us | 42.72 us |  1.00 |    0.00 | 0.6104 |  16.88 KB |        1.00 |

Fixes #2615

Commit:
Don't always test string claims for Date Time

Cannot remove APIs in in these internal methods as they are friends

There are serveral claim types that should never
be a datetime value. This changes introduces
a list of claim types that will not be tested if they
are a Date Time format. This list of claim types is
not exhaustive.

Added test data for a test token with more claims
and a benchmark using this data.

keegan-caruso commented on src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs:
```diff
@@ -695,6 +706,98 @@ internal static string ReadStringOrNumberAsString(ref Utf8JsonReader reader, str
+        /// sourced from expected Entra V1 and V2 claims, OpenID Connect claims, and a selection of
+        /// restricted claim names.
+        /// </summary>
+        public static HashSet<string> KnownNonDateTimesClaimTypes = new(StringComparer.Ordinal)
```
There may be a better type to put this on, open to suggestions

brentschmaltz replied:
I think this can be internal.
The switch will handle backcompat.

keegan-caruso replied:
I'll put this behind an internal method, no need for the Collection to be exposed

JoshLozensky replied:
are these values not already available as constants somewhere? do we really need to make all of these strings?

keegan-caruso replied:
These are all from external sources, I think it is more important to be able to read the values exactly in this case vs having them in constants elsewhere. I'd agree if they got uses elsewhere.

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs:
```diff
@@ -695,6 +706,98 @@ internal static string ReadStringOrNumberAsString(ref Utf8JsonReader reader, str
+
+        public static bool TryAllStringClaimsAsDateTime()
+        {
+            return (AppContext.TryGetSwitch(TryToCreateDateTimeClaimsSwitch, out bool tryAsDateTime) && tryAsDateTime);
```
what is the default?

keegan-caruso replied:
Default is the new behavior

brentschmaltz replied:
I take it that means that by default we will not attempt to convert and of the claims in KnownNonDateClaimTypes.

keegan-caruso replied:
Yes exactly.

Commit:
Prefer to expose method over collection

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs:
```diff
@@ -695,6 +706,109 @@ internal static string ReadStringOrNumberAsString(ref Utf8JsonReader reader, str
+        /// sourced from expected Entra V1 and V2 claims, OpenID Connect claims, and a selection of
+        /// restricted claim names.
+        /// </summary>
+        private static HashSet<string> KnownNonDateTimesClaimTypes = new(StringComparer.Ordinal)
```
We may want to think about header claims.
[JWT:](https://datatracker.ietf.org/doc/html/rfc7519#section-5) , cty, typ, [JWS](https://datatracker.ietf.org/doc/html/rfc7515#section-4.1) alg, crit, jku, jwk, kid, x5u, x5c, x5t, x5t#256, typ  [JWE](https://datatracker.ietf.org/doc/html/rfc7516#section-4.1) enc, zip

Question is do we lump payload and header in the same switch?
I think it's probably OK to have one list, we can always expose this later and give users complete control.

keegan-caruso replied:
added

brentschmaltz reviewed (changes requested):
Asking to add the additional header claim values.

Commit:
Add JWT headers to known non datetimes values

pmaytak commented on src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.cs:
```diff
@@ -599,8 +600,18 @@ internal static IEnumerable<SecurityKey> ConcatSigningKeys(TokenValidationParame
+            return GetStringClaimValueType(string.Empty, str);
+        }
+
+        internal static string GetStringClaimValueType(string claimType, string str)
```
Revert the order to match the above overload.
```suggestion
        internal static string GetStringClaimValueType(string str, string claimType)
```

pmaytak commented on src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs:
```diff
@@ -129,6 +129,14 @@ public static JsonElement CreateJsonElement(string json)
+        /// <remarks>
+        /// <paramref name="claimType"/> is not considered on recursive calls.
+        /// </remarks>
+        internal static object CreateObjectFromJsonElement(string claimType, JsonElement jsonElement, int currentDepth)
```
Same comment. Change the param order to match above.

pmaytak commented on src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs:
```diff
@@ -695,6 +706,122 @@ internal static string ReadStringOrNumberAsString(ref Utf8JsonReader reader, str
+        /// sourced from expected Entra V1 and V2 claims, OpenID Connect claims, and a selection of
+        /// restricted claim names.
+        /// </summary>
+        private static HashSet<string> KnownNonDateTimesClaimTypes = new(StringComparer.Ordinal)
```
nit:
```suggestion
        private static HashSet<string> KnownNonDateTimeClaimTypes = new(StringComparer.Ordinal)
```

pmaytak commented on src/System.IdentityModel.Tokens.Jwt/JwtPayload.cs:
```diff
@@ -664,15 +664,15 @@ internal void AddDictionaryClaims(IDictionary<string, object> claimsCollection)
         }
 
-        internal static string GetClaimValueType(object value)
+        private static string GetClaimValueType(string claimType, object value)
```
Safe to make private?

keegan-caruso replied:
This one yes, it wasn't used outside of this method.

pmaytak commented on src/System.IdentityModel.Tokens.Jwt/JwtPayload.cs:
```diff
@@ -664,15 +664,15 @@ internal void AddDictionaryClaims(IDictionary<string, object> claimsCollection)
         }
 
-        internal static string GetClaimValueType(object value)
+        private static string GetClaimValueType(string claimType, object value)
```
Is there a reason for adding new params to the begging? I think per guidelines, they're usually added to the end.

keegan-caruso replied:
Closer to the signature of the other methods.

pmaytak commented on src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs:
```diff
@@ -695,6 +706,122 @@ internal static string ReadStringOrNumberAsString(ref Utf8JsonReader reader, str
             return retVal;
         }
 
+        internal const string TryToCreateDateTimeClaimsSwitch = "Switch.Microsoft.IdentityModel.TryAllStringClaimsAsDateTime";
```
Do we really need a switch for this? Or is this more of a safety measure to disable quickly if there's some unintended behaivor?

keegan-caruso replied:
exactly. It is unlikely, but we wanted an opt out.

Commit:
Review feedback

- Reorder params in new methods
- Rename KnownNonDateTimesClaimTypes to s_knownNonDateTimeClaimTypes
